README.md
# fourclojure-soln

practice with [4clojure](http://www.4clojure.com/problems)

## Personal Usage

1. Select problem 
2. Add sample cases to `core_test.clj` with name of problem
3. Add solution to `core.clj` with name of problem 
4. Run `lein test`



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2019-09-19
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2019-09-19
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/fourclojure-soln/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/fourclojure-soln/compare/0.1.0...0.1.1


test/fourclojure_soln/core_test.clj
(ns fourclojure-soln.core-test
  (:require [clojure.test :refer :all]
            [fourclojure-soln.core :refer :all]))

;; #19 
(deftest lastelm-test
  (testing "Last Element"
    (is (= (lastelm [1 2 3 4 5]) 5))
    (is (= (lastelm '(5 4 3)) 3))
    (is (= (lastelm ["b" "c" "d"] )"d"))
    ))

;; #20
(deftest penultimate-test
  (testing "Penultimate Element"
    (is (= (penultimate (list 1 2 3 4 5)) 4))
    (is (= (penultimate ["a" "b" "c"]) "b"))
    (is (= (penultimate [[1 2] [3 4]]) [1 2]))
    ))

;; #21
(deftest nth-elm-test
  (testing "Nth Element"
    (is (= (nth-elm '(4 5 6 7) 2) 6))
    (is (= (nth-elm [:a :b :c] 0) :a))
    (is (= (nth-elm [1 2 3 4] 1) 2))
    (is (= (nth-elm '([1 2] [3 4] [5 6]) 2) [5 6]))
    ))

;; #22
(deftest my-count-test
  (testing "Count a Sequence"
    (is (= (my-count '(1 2 3 3 1)) 5))
    (is (= (my-count "Hello World") 11))
    (is (= (my-count [[1 2] [3 4] [5 6]]) 3))
    (is (= (my-count '(13)) 1))
    (is (= (my-count '(:a :b :c)) 3))
    ))

;; #23
(deftest my-reverse-test
  (testing "Reverse a Sequence"
    (is (= (my-reverse [1 2 3 4 5]) [5 4 3 2 1]))
    (is (= (my-reverse (sorted-set 5 7 2 7)) '(7 5 2)))
    (is (= (my-reverse [[1 2][3 4][5 6]]) [[5 6][3 4][1 2]]))
    ))

;; #24
(deftest sum-all-test
  (testing "Reverse a Sequence"
    (is (= (sum-all [1 2 3]) 6))
    (is (= (sum-all (list 0 -2 5 5)) 8))
    (is (= (sum-all #{4 2 1}) 7))
    (is (= (sum-all '(0 0 -1)) -1))
    (is (= (sum-all '(1 10 3)) 14))
    ))

;; #25
(deftest find-odd-test
  (testing "Find the odd numbers"
    (is (= (find-odd #{1 2 3 4 5}) '(1 3 5)))
    (is (= (find-odd [4 2 1 6]) '(1)))
    (is (= (find-odd [2 2 4 6]) '()))
    (is (= (find-odd [1 1 1 3]) '(1 1 1 3)))
    ))

;; #26 
(deftest fib-test
  (testing "Fibonacci Sequence"
    (is (= (fiblist 3) '(1 1 2)))
    (is (= (fiblist 6) '(1 1 2 3 5 8)))
    (is (= (fiblist 8) '(1 1 2 3 5 8 13 21)))

))

;; #27
(deftest palindrome-test
  (testing "Palindrome Detector"
    (is (false? (palindrome? '(1 2 3 4 5))))
    (is (true? (palindrome? "racecar")))
    (is (true? (palindrome? [:foo :bar :foo])))
    (is (true? (palindrome? '(1 1 3 3 1 1))))
    (is (false? (palindrome? '(:a :b :c))))
))

;; 29
(deftest getcaps-test
  (testing "Get the Caps"
    (is (= (getcaps "HeLlO, WoRlD!") "HLOWRD"))
    (is (empty? (getcaps "nothing")))
    (is (= (getcaps "$#A(*&987Zf") "AZ"))
))

;; 28
(deftest myflatten-test
  (testing "Flatten a Sequence"
    (is (= (my-flatten '((1 2) 3 [4 [5 6]])) '(1 2 3 4 5 6)))
    (is (= (my-flatten ["a" ["b"] "c"]) '("a" "b" "c")))
    (is (= (my-flatten '((((:a))))) '(:a)))
  )
)

;; #30 
(deftest compress-test
  (testing "Compress a Sequence"
    (is (= (apply str (compress "Leeeeeerrroyyy")) "Leroy"))
    (is (= (compress [1 1 2 3 3 3 2 2 3]) '(1 2 3 2 3)))
    (is (= (compress [[1 2] [1 2] [3 4] [1 2]]) '([1 2] [3 4] [1 2])))
  )
)

;; #31
(deftest packseq-test
  (testing "Pack a Sequence"
    (is (= (packseq [1 1 2 1 1 1 3 3]) '((1 1) (2) (1 1 1) (3 3))))
    (is (= (packseq [:a :a :b :b :c]) '((:a :a) (:b :b) (:c))))
    (is (= (packseq [[1 2] [1 2] [3 4]]) '(([1 2] [1 2]) ([3 4]))))
    ))
    
;; #32
(deftest dupseq-test
  (testing "Duplicate a Sequence"
    (is (= (dupseq [1 2 3]) '(1 1 2 2 3 3)))
    (is (= (dupseq [:a :a :b :b]) '(:a :a :a :a :b :b :b :b)))
    (is (= (dupseq [[1 2] [3 4]]) '([1 2] [1 2] [3 4] [3 4])))
    ))

;; #33
(deftest repseq-test
  (testing "Replicate a Sequence"
    (is (= (repseq [1 2 3] 2) '(1 1 2 2 3 3)))
    (is (= (repseq [:a :b] 4) '(:a :a :a :a :b :b :b :b)))
    (is (= (repseq [4 5 6] 1) '(4 5 6)))
    (is (= (repseq [[1 2] [3 4]] 2) '([1 2] [1 2] [3 4] [3 4])))
    (is (= (repseq [44 33] 2) [44 44 33 33]))
  )
)

;; #34
(deftest myrange-test
  (testing "Implement Range"
    (is (= (myrange 1 4) '(1 2 3)))
    (is (= (myrange -2 2) '(-2 -1 0 1)))
    (is (= (myrange 5 8) '(5 6 7)))
  )
)

;; #38
(deftest mymax-test
  (testing "Maximum Value"
    (is (= (mymax 1 8 3 4) 8))
    (is (= (mymax 30 20) 30))
    (is (= (mymax 45 67 11) 67))
  )
)

src/fourclojure_soln/core.clj
(ns fourclojure-soln.core)

(defn foo
  "I don't do a whole lot."
  [x]
  (println x "Hello, World!"))

;; #19 Last Element
(defn lastelm 
  "returns the last element in a sequence"
  [[frst & rst :as elms]]
  (cond 
   	(= (count elms) 0) nil
  	(= (count elms) 1) frst
	  :else (lastelm rst)
))

;; #20 Penultimate Element
(defn penultimate 
  "returns second to last element from sequence"
  [elms]
  (if (= 2 (count elms)) (first elms) (penultimate (rest elms))))

;; #21 Nth Element
(defn nth-elm 
  "returns nth element from sequence"
  [elms, n]
  (if (= 0 n) (first elms) (nth-elm (rest elms) ( - n 1))))

;; #22 Count a Sequence
(defn my-count 
  "returns total number of elements in sequence"
  [elms]
  (if (empty? elms) 0 (+ 1 (my-count (rest elms)))))

;; #23 Reverse a Sequence
(defn my-reverse 
  "returns reverse of sequence"
  [elms]
  (reduce conj () elms))

;; #24 Sum It All Up
(defn sum-all 
  "returns sum of seq of numbers"
  [elms]
  (reduce + 0 elms))

;; #25 Find the odd numbers
(defn find-odd 
  "returns odd numbers in a sequence"
  [elms]
  (filter odd? elms))


;; #26 Fibonacci Sequence

;; Helper
(defn fib 
  "returns nth fibonacci number"
  [n]
  (cond 
    (<= n 2) 1
    :else (+ (fib (- n 1)) (fib (- n 2)))
  )
)
;; Fib Sequence
(defn fiblist 
  "returns first n fibonacci numbers"
  [n]
  (map fib (range 1 (inc n)))
)


;; #27 Palindrome Detector
(defn palindrome? 
  "returns true of given sequence is palindrome"
  [elms]
  (if (or (empty? elms) (= 1 (count elms))) true
    (and (= (first elms)(last elms)) 
      (palindrome? (subvec (vec elms) 1 (- (count elms ) 1))
    )) 
  )
)

;; #28 Flatten a Sequence
(defn my-flatten 
  "removes nested sequences to form one dimensional sequence"
  [[frst & rst :as elms]]
  (cond
    (empty? elms) nil
    (not (coll? frst)) (cons frst (my-flatten rst))
    :else (concat (my-flatten frst) (my-flatten rst))
  )
)

;; #29 Get the Caps
(defn getcaps 
  "returns new string containing only capital letters"
  [string]
  (apply str (filter #(Character/isUpperCase %) string))
  )

;; #30 Compress a Sequence
(defn compress 
  "removes consecutive duplicates from sequence"
  [elms]
  (cond 
    (or (empty? elms)(= 1 (count elms))) elms
    (= (first elms)(first (rest elms)))  (compress (rest elms))
    :else (concat (list (first elms)) (compress (rest elms)))
))

	

;; #31 Pack a Sequence 

;; Helper 
(defn numrep 
  "return number of consecutive duplicates from the start of a sequence"
  [[frst & others]]
  (cond
    (nil? frst) 0
    (empty? others) 1
    (= frst (first others)) (+ 1 (numrep others))
    :else 1
  ))

;; Pack a Sequence
(defn packseq 
  "return sequence with consecutive duplicates into sublists"
  [vec]
  (if (empty? vec) () 
  (concat (list (repeat (numrep vec) (first vec))) (packseq (subvec vec (numrep vec))))
  ))

;; alt solution: (partition-by identity?)


;; #32 Duplicate a Sequence
(defn dupseq 
  "returns sequence with each element duplicated"
  [elms]
  (apply concat (map #(list % %) elms))
  )

;; #33 Replicate a Sequence
(defn repseq 
  "returns sequence with each element repeated n times"
  [elms, n]
  (apply concat (map #(repeat n %) elms))
)

;; #34 Implement Range
(defn myrange [start end]
  "creates list of all integers in given range"
  (if (> start (dec end)) () (cons start (myrange (inc start) end)))
)

;; #38 Maximum Value
(defn mymax 
  "takes variable number of parameters and returns maximum value"
  [& nums]
  (reduce #(if (< %1 %2) %2 %1) nums)
)